specs/PushNotifications/ToastNotifications-spec.md

```diff
@@ -0,0 +1,567 @@
+NotificationBuilder model in the near future for WinAppSdk.
+
+# Description
+
+At a high level, we need to provide a way for all Win32 applications to display app notifications irrespective
+of their app type. This includes unpackaged apps and packaged win32 (MSIX Desktop Bridge, MSIX
+Win32App, Sparse Signed Packages). Moreover, all scenarios should adhere to OS resource
+management policies like Power Saver, Network Attribution (amount of data an app uses), enterprise
+group policies, etc. The WinAppSDK will abstract away the complexities of
+dealing with Notification delivery and related activations as much as possible freeing the developer
+to focus on other app related challenges.
+
+We will prioritize the following feature set for Notifications in Windows App SDK:
+
+-   App Notification Registrations for Unpackaged apps.
+-   App Notification Registrations for Packaged apps
+-   Cloud sourced App Notifications for Unpackaged Win32 apps.
+-   Cloud sourced App Notifications for Packaged Win32 apps.
+-   Local App Notification delivery and Create/Update/Delete support for Packaged Win32 apps.
+-   Local App Notification delivery and Create/Update/Delete support for unpackaged Win32 apps.
+
+The following feature set will not be prioritized for the future include:
+
+-   Object Model to compose payloads without having to rely on Xml templates.
+-   Support for categorizing App Notifications by Collection.
+
+# Examples
+
+## Packaged App Notification Registration
+
+This scenario is specifically geared towards Packaged apps that need to send an app notification. An app
+Developer Registers their app for notifications and sets it up for background\foreground
+activation. Adding a AppNotificationActivated event handler signals that the app is ready to accept and handle
+an action in the foreground where the user clicks on a Notification in the action centre or a UI element in
+the payload body (like a Button). The developer can inturn run some business logic in their app in
+response to the UI Action.
+
+```cpp
+int main()
+{
+  auto activationInfo = AppNotificationActivationInfo::CreateFromActivationGuid(winrt::guid(c_comServerId));
+    AppNotificationManager::Default().RegisterActivator(activationInfo);
+
+    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();
+    auto kind = args.Kind();
+
+    if (kind == ExtendedActivationKind::Launch)
+    {
+        // App is launched in Foreground. So intercept Notification activators via Foreground event
+        const auto token = AppNotificationManager::Default().AppNotificationActivated([](const auto&, const AppNotificationActivatedEventArgs& notificationActivatedEventArgs)
+        {
+                ProcessNotificationArgs(notificationActivatedEventArgs);
+        });
+
+        // App does Foreground Stuff Here
+
+        // Cleanup
+        AppNotificationManager::Default().AppNotificationActivated(token);
+    }
+    else if (kind == ExtendedActivationKind::AppNotification)
```

Comment:
is this the background case? only executing if the app is not already running? does this use the -Embedding command line to detect the COM activation?

Reply:
Correct. This is specifically for the background case. WinAppSDK will parse the command line for -Embedding to detect the activator args.

Reply:
1. The app will hook up the event handle before registering.

The notification subsystem does not guarantee activation ordering, so there is no "new" race between the event and the activation args.

Why not just use the event for all cases?

DISCUSSION: If the developer never calls Register, then all activations start a new process and go through the "case AppNotification" path.

When the app calls Register, the AppNotificationManager checks if there is any handler registered for AppNotificationActivated. if not, then it assumes the app wants a new activation for every notification.

RECOMMEND: Take offline to reconcile against the principle that registering an event handler does not alter behavior.